### PR TITLE
Add ReadDigitalWaveforms to NiDAQmxService

### PR DESCRIPTION
**Pull request by mikeprosserni**
### What does this Pull Request accomplish?

-Add ReadDigitalWaveforms method to the NiDAQmxService

### Why should this Pull Request be merged?

[AB#3424630](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/3424630)

### What testing has been done?

I added four tests for ReadDigitalWaveforms:

image

I also update the [proof of concept in nidaqmx-python](https://github.com/ni/nidaqmx-python/compare/users/mprosser/grpc-waveform-proof-of-concept?expand=1) to use this new ReadDigitalWaveforms RPC, and now the grpc versions of the related tests are passing:

image

(This demonstrates that all the multi-channel/multi-line/port handling is working as expected.)

### CODE REVIEW

File: source/custom/nidaqmx_service.custom.cpp [hunk 1]
```diff
@@ -93,6 +94,79 @@ int32 CVICALLBACK SetWfmAttrCallback(
...
+    const void* value,
+    const uInt32 value_size_in_bytes,
+    void* callback_data)
+{
+  return SetWfmAttrCallbackTemplated<RepeatedPtrField<::ni::protobuf::types::DigitalWaveform>>(
+    channel_index, attribute_name, attribute_type, value, value_size_in_bytes, callback_data);
+}
+
+template<typename RequestType>
+int32 ParseWaveformAttributeMode(const RequestType* request)
```
Comment (bkeryan) [inline source/custom/nidaqmx_service.custom.cpp]: Parsing usually refers to extracting data from a string. Consider calling this "GetWaveformAttributeMode".

File: source/custom/nidaqmx_service.custom.cpp [hunk 2]
```diff
@@ -93,6 +94,79 @@ int32 CVICALLBACK SetWfmAttrCallback(
...
+      return static_cast<int32>(request->waveform_attribute_mode_raw());
+    }
+    case RequestType::WaveformAttributeModeEnumCase::WAVEFORM_ATTRIBUTE_MODE_ENUM_NOT_SET: {
+      return 0;
+    }
+  }
+  return 0;
+}
+
+void SetupTimingArrays(
```
Comment (bkeryan) [inline source/custom/nidaqmx_service.custom.cpp]: "setup" is a noun.
"set up" is a verb.
"initialize" is more commonly used to describe this.

File: source/tests/system/nidaqmx_driver_api_tests.cpp
Comment (bkeryan) [inline source/tests/system/nidaqmx_driver_api_tests.cpp]: > I also update the [proof of concept in nidaqmx-python](https://github.com/ni/nidaqmx-python/compare/users/mprosser/grpc-waveform-proof-of-concept?expand=1) to use this new ReadDigitalWaveforms RPC, and now the grpc versions of the related tests are passing:

Good, I think it's more expedient to rely on the Python tests for the corner cases.

BTW, I'm still concerned about AB#3178052. We may have the same ordering inconsistency in both Python and gRPC.

File: source/custom/nidaqmx_service.custom.cpp [hunk 3]
```diff
@@ -93,6 +94,79 @@ int32 CVICALLBACK SetWfmAttrCallback(
   }
 }
 
+int32 CVICALLBACK SetAnalogWfmAttrCallback(
```
[the remark this answers is not in the document]
Comment (mikeprosserni) [reply source/custom/nidaqmx_service.custom.cpp]: Updated to use `callback_ptr = &SetWfmAttrCallbackTemplated<RepeatedPtrField<DoubleAnalogWaveform>>;` and `callback_ptr = &SetWfmAttrCallbackTemplated<RepeatedPtrField<DigitalWaveform>>;` instead of the wrappers.

File: source/custom/nidaqmx_service.custom.cpp (hunk 2 above)
Comment (mikeprosserni) [reply source/custom/nidaqmx_service.custom.cpp]: changed to `TimingData` constructor

File: source/tests/system/nidaqmx_driver_api_tests.cpp
Comment (mikeprosserni) [reply source/tests/system/nidaqmx_driver_api_tests.cpp]: Yes, we definitely have the same ordering inconsistency. The `test___digital_multi_channel_lines_and_port_reader___read_waveforms___returns_valid_waveforms` test ensures the behavior is the same for both (since the test checks both _library_interpreter.py and _grpc_interpreter.py), and it includes this:
```
    # Note, the data on the port's waveform is MSB instead of LSB because of bug AB#3178052
    # When that bug is fixed, these asserts should be updated
    assert _get_waveform_data(waveforms[3]) == [0, 128, 64, 192, 32, 160, 96, 224, 16, 144]
```

I'm planning to work on AB#3178052 once I finish adding these waveform functions. With them all in place, it may be easer than when Joe was trying to work on it back in September.

File: source/custom/nidaqmx_service.custom.cpp [hunk 4]
```diff
@@ -12,16 +12,37 @@ using nidevice_grpc::converters::convert_dot_net_ticks_to_precision_timestamp;
...
+    t0_ptr = t0_array.data();
+    dt_ptr = dt_array.data();
+    timing_array_size = num_channels;
+  }
+};
+
 // Universal callback function for setting waveform attributes (both analog and digital)
 template<typename WaveformCollectionType>
-int32 CVICALLBACK SetWfmAttrCallbackTemplated(
+static int32 CVICALLBACK SetWfmAttrCallbackTemplated(
```
Comment (bkeryan) [inline source/custom/nidaqmx_service.custom.cpp]: Remove "templated"
```suggestion
static int32 CVICALLBACK SetWfmAttrCallback(
```

File: source/custom/nidaqmx_service.custom.cpp [hunk 5]
```diff
@@ -12,16 +12,37 @@ using nidevice_grpc::converters::convert_dot_net_ticks_to_precision_timestamp;
...
+using ::ni::protobuf::types::DigitalWaveform;
 
 // Returns true if it's safe to use outputs of a method with the given status.
 inline bool status_ok(int32 status)
 {
   return status >= 0;
 }
 
+// Encapsulates timing arrays and pointers for waveform operations
+struct TimingData {
```
[the remark this answers is not in the document]
Comment (mikeprosserni) [reply source/custom/nidaqmx_service.custom.cpp]: moved the helper classes and functions to anonymous namespace

File: source/custom/nidaqmx_service.custom.cpp (hunk 4 above)
Comment (mikeprosserni) [reply source/custom/nidaqmx_service.custom.cpp]: renamed

**Review by reckenro — APPROVED** [review]
